pkg/app/api/grpcapi/piped_api.go

```diff
@@ -963,6 +967,129 @@ func (a *PipedAPI) ReportUnregisteredApplicationConfigurations(ctx context.Conte
 	return nil, status.Errorf(codes.Unimplemented, "ReportUnregisteredApplicationConfigurations is not implemented yet")
 }
 
+// CreateDeploymentChain creates a new deployment chain object and all required commands to
+// trigger deployment for applications in the chain.
+func (a *PipedAPI) CreateDeploymentChain(ctx context.Context, req *pipedservice.CreateDeploymentChainRequest) (*pipedservice.CreateDeploymentChainResponse, error) {
+	projectID, _, _, err := rpcauth.ExtractPipedToken(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	buildChainNodes := func(matcher *pipedservice.CreateDeploymentChainRequest_ApplicationMatcher) ([]*model.ChainNode, []*model.Application, error) {
+		filters := []datastore.ListFilter{
+			{
+				Field:    "ProjectId",
+				Operator: datastore.OperatorEqual,
+				Value:    projectID,
+			},
+		}
+
+		if matcher.Name != "" {
+			filters = append(filters, datastore.ListFilter{
+				Field:    "Name",
+				Operator: datastore.OperatorEqual,
+				Value:    matcher.Name,
+			})
+		}
+
+		// TODO: Support find node apps by appKind and appLabels.
+
+		apps, _, err := a.applicationStore.ListApplications(ctx, datastore.ListOptions{
+			Filters: filters,
+		})
+		if err != nil {
+			return nil, nil, err
+		}
+
+		nodes := make([]*model.ChainNode, 0, len(apps))
+		for _, app := range apps {
+			nodes = append(nodes, &model.ChainNode{
+				ApplicationRef: &model.ChainApplicationRef{
+					ApplicationId:   app.Id,
+					ApplicationName: app.Name,
+				},
+			})
+		}
+
+		return nodes, apps, nil
+	}
+
+	chainBlocks := make([]*model.ChainBlock, 0, len(req.Matchers)+1)
+	// Add the first deployment which created by piped as the first block of the chain.
+	chainBlocks[0] = &model.ChainBlock{
```

Comment:
Have to use `append` to avoid panic because the size of `chainBlocks` was initialized by 0.

Reply:
Nice catch, addressed by https://github.com/pipe-cd/pipe/pull/2815/commits/8790eda637f06196a46a23b46f044bc95d1bb458